src/common/config.h

```diff
@@ -2,84 +2,161 @@
  * Copyright 2014-2019 by Contributors
  * \file config.h
  * \brief helper class to load in configures from file
- * \author Haoda Fu
+ * \author Haoda Fu, Hyunsu Cho
  */
 #ifndef XGBOOST_COMMON_CONFIG_H_
 #define XGBOOST_COMMON_CONFIG_H_
 
+#include <dmlc/logging.h>
 #include <cstdio>
-#include <cstring>
 #include <string>
 #include <istream>
 #include <fstream>
 #include <vector>
+#include <regex>
+#include <iterator>
 #include <utility>
 
 namespace xgboost {
 namespace common {
 /*!
  * \brief Implementation of config reader
  */
-class ConfigParse {
+class ConfigParser {
  public:
   /*!
-   * \brief constructor
-   * \param cfgFileName name of configure file
+   * \brief Constructor for INI-style configuration parser
+   * \param path path to configuration file
    */
-  explicit ConfigParse(const std::string &cfgFileName) {
-    fi_.open(cfgFileName);
-    if (fi_.fail()) {
-      LOG(FATAL) << "cannot open file " << cfgFileName;
-    }
+  explicit ConfigParser(const std::string& path)
```

Comment:
please document the grammar. it is a bit difficult to grok all this without poring over this in greater detail.

Reply:
I will go ahead and write a doc for the grammar.